cpp/src/arrow/util/list_util.cc

```diff
@@ -0,0 +1,285 @@
+
+  return ArrayData::Make(std::make_shared<DestListViewType>(list_type.value_type()),
+                         list_data->length, std::move(buffers),
+                         {list_data->child_data[0]}, list_data->null_count,
+                         list_data->offset);
+}
+
+template <typename DestListType, typename SrcListViewType>
+Result<std::shared_ptr<ArrayData>> ListFromListViewImpl(
+    const std::shared_ptr<ArrayData>& list_view_data, MemoryPool* pool) {
+  static_assert(
+      std::is_same<typename SrcListViewType::offset_type,
+                   typename DestListType::offset_type>::value,
+      "Offset types between list type and list-view type are expected to match");
+  using offset_type = typename DestListType::offset_type;
+  using ListBuilderType = typename TypeTraits<DestListType>::BuilderType;
+
+  const auto& list_view_type =
+      checked_cast<const SrcListViewType&>(*list_view_data->type);
+  const auto& value_type = list_view_type.value_type();
+  const auto list_type = std::make_shared<DestListType>(value_type);
+
+  ARROW_ASSIGN_OR_RAISE(std::shared_ptr<ArrayBuilder> value_builder,
+                        MakeBuilder(value_type, pool));
+  auto list_builder = std::make_shared<ListBuilderType>(pool, value_builder, list_type);
+  RETURN_NOT_OK(list_builder->Reserve(list_view_data->length));
+
+  ArraySpan values{*list_view_data->child_data[0]};
+  const auto* in_validity_bitmap = list_view_data->GetValues<uint8_t>(0);
+  const auto* in_offsets = list_view_data->GetValues<offset_type>(1);
+  const auto* in_sizes = list_view_data->GetValues<offset_type>(2);
+  for (int64_t i = 0; i < list_view_data->length; ++i) {
+    const bool is_valid =
+        !in_validity_bitmap ||
+        bit_util::GetBit(in_validity_bitmap, list_view_data->offset + i);
+    const int64_t size = is_valid ? in_sizes[i] : 0;
+    RETURN_NOT_OK(list_builder->Append(is_valid, size));
+    RETURN_NOT_OK(value_builder->AppendArraySlice(values, in_offsets[i], size));
+  }
+  std::shared_ptr<ArrayData> list_array_data;
+  RETURN_NOT_OK(list_builder->FinishInternal(&list_array_data));
+  return list_array_data;
+}
+
+}  // namespace
+
+Result<std::pair<int64_t, int64_t>> RangeOfValuesUsed(const ArraySpan& input) {
+  switch (input.type->id()) {
+    case Type::LIST:
+      return RangeOfValuesUsedByList<ListType::offset_type>(input);
+    case Type::MAP:
+      return RangeOfValuesUsedByList<MapType::offset_type>(input);
+    case Type::LARGE_LIST:
+      return RangeOfValuesUsedByList<LargeListType::offset_type>(input);
+    case Type::LIST_VIEW:
+      return RangeOfValuesUsedByListView<ListViewType::offset_type>(input);
+    case Type::LARGE_LIST_VIEW:
+      return RangeOfValuesUsedByListView<LargeListViewType::offset_type>(input);
+    default:
+      break;
+  }
+  DCHECK(!is_var_length_list_like(*input.type));
+  return Status::Invalid("RangeOfValuesUsed: input is not a var-length list-like array");
```

Comment:
Make this `Status::TypeError` since it's a type error.

Reply:
Done.